app/lib/account/backend.dart

```diff
@@ -0,0 +1,195 @@
+  final DatastoreDB _db;
+  final _defaultAuthProvider =
+      GoogleOauth2AuthProvider('dartlang-pub-1', _pubAudience);
+
+  AccountBackend(this._db);
+
+  Future close() async {
+    await _defaultAuthProvider.close();
+    await _pool.close();
+  }
+
+  /// Returns the `User` entry that is associated with the [accessToken].
+  ///
+  /// When no entry exists in Datastore, this method will create a new one.
+  ///
+  /// When the authenticated e-mail of the user changes, the email field will
+  /// be updated to the latest one, and the `historicalEmails` field contains
+  /// all of the earlier and current values.
+  ///
+  /// The method returns null if the access token is invalid.
+  Future<User> lookupOrCreateUser(String accessToken) async {
+    final auth = await _defaultAuthProvider.tryAuthenticate(accessToken);
+    if (auth == null) {
+      return null;
+    }
+    final compositeId = auth.compositeId;
+    final userKey = _db.emptyKey.append(User, id: compositeId);
+
+    final accessTime = DateTime.now().toUtc();
+    final user = await _retry.retry(() async {
+      User user = (await _db.lookup<User>([userKey])).single;
+      if (user == null) {
+        // create user
+        user = User()
+          ..parentKey = _db.emptyKey
+          ..id = compositeId
+          ..authProvider = auth.provider
+          ..authUserId = auth.userId
+          ..email = auth.email
+          ..historicalEmails = <String>[auth.email]
+          ..created = accessTime
+          ..updated = accessTime;
+        _db.commit(inserts: [user]);
+      } else {
+        // update user if e-mail has been changed
+        if (user.email != auth.email) {
+          await _db.withTransaction((tx) async {
+            user = (await _db.lookup<User>([userKey])).single;
+            user.updateEmail(auth.email);
+            user.updated = accessTime;
+            tx.queueMutations(inserts: [user]);
+            await tx.commit();
+          });
+        }
+      }
+      return user;
+    });
+
+    if (user == null) {
+      return null;
+    }
+
+    // Future not awaited: updating the last accessed field does not need to
+    // block the request serving process.
+    _pool.withResource(() async {
+      _retry.retry(() async {
+        await _db.withTransaction((tx) async {
+          final user = (await _db.lookup<User>([userKey])).single;
+          if (user.lastAccess == null || user.lastAccess.isBefore(accessTime)) {
+            user.lastAccess = accessTime;
+            tx.queueMutations(inserts: [user]);
+            await tx.commit();
+          } else {
+            await tx.rollback();
+          }
+        });
+      });
+    });
+
+    return user;
+  }
+}
+
+class AuthResult {
+  final String provider;
+  final String userId;
+  final String email;
+
+  AuthResult(this.provider, this.userId, this.email);
+
+  String get compositeId => '$provider:$userId';
+}
+
+/// Authenticates access tokens.
+abstract class AuthProvider {
```

Comment:
This seems like an unnecessary abstraction...

Reply:
I think the abstraction could be still useful for local development and/or tests.